### PR TITLE
Add testing and fix issues for weights_only load for LRScheduler

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is described in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.Counter`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
  Fixed the above in a BC-compatible manner by adding `OneCyclicLR._anneal_func_type` as a string attribute and removing `OneCyclicLR.anneal_func`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123775

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -1761,6 +1761,18 @@ def _format_param(self, name, optimizer, param):
         else:
             return [param] * len(optimizer.param_groups)
 
+
+    def state_dict(self):
```
Comment (mikaylagawarecki) [inline torch/optim/lr_scheduler.py:1765]: Use versioning here to ensure old `state_dict` can be loaded

Comment (albanD) [reply torch/optim/lr_scheduler.py:1765]: I think a better refactor is:
- Add a `self._anneal_func_type` that is the string from the user
- Add a `self._anneal_func` method on the class that will check the _anneal_func_type field and call the appropriate method or will call self.anneal_func when that exists (for BC)

That way you don't need to override state_dict/load_state_dict

Comment (mikaylagawarecki) [reply torch/optim/lr_scheduler.py]: Made the suggested fix

File: test/optim/test_lrscheduler.py [hunk 2]
```diff
@@ -2285,6 +2286,41 @@ def test_lr_scheduler_verbose_deprecation_warning(self, LRClass):
             LRClass(self.opt)
 
 
+    @parametrize("LRClass", [
```
Comment (albanD) [inline test/optim/test_lrscheduler.py:2289]: Can you make a one off test to make sure you can load an old state dict that is hardcoded? Just to make sure the BC handling is working as expected?

**Review by albanD — APPROVED** [review]
SGTM!
Just a small test to exercise the BC handling code

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**